File: src/VMS.Modules.Vehicles/Models/VehicleModels.cs
using VMS.Shared.Authorization;
using VMS.Shared.Pagination;

namespace VMS.Modules.Vehicles.Models;

/// <summary>What is sent to save a vehicle's identity, technical and operational fields (FSD §16). Everything else about a vehicle has its own action.</summary>
public class VehicleInput
{
    public string RegistrationNo { get; set; } = string.Empty;
    public int? RegistrationCityId { get; set; }
    public string? ChassisNo { get; set; }
    public string? EngineNo { get; set; }
    /// <summary>A value of the Vehicle Type list.</summary>
    public int VehicleTypeId { get; set; }
    /// <summary>A value of the Make list.</summary>
    public int MakeId { get; set; }
    public string Model { get; set; } = string.Empty;
    public int? ManufacturingYear { get; set; }
    public string? Colour { get; set; }

    /// <summary>Diesel, Petrol, CNG, LPG, Hybrid or Electric.</summary>
    public string FuelType { get; set; } = string.Empty;
    public decimal? TankCapacity { get; set; }
    /// <summary>Required for a Truck, Trailer, Tanker or Prime Mover.</summary>
    public decimal? LoadCapacity { get; set; }
    /// <summary>Tonne, Kg, Litre, CFT or Passengers. Required when a load capacity is entered.</summary>
    public string? CapacityUnit { get; set; }
    public int? AxleConfigurationId { get; set; }
    public int? BodyTypeId { get; set; }
    public int? TyreCount { get; set; }
    public decimal? Gvw { get; set; }

    public Guid? BranchId { get; set; }
    public int? OpeningOdometer { get; set; }
    /// <summary>Required when an opening odometer is entered. Not in the future.</summary>
    public DateOnly? OpeningOdometerDate { get; set; }
    /// <summary>A Driver partner. On a vehicle already in the fleet the default driver is changed with the assign action, not here.</summary>
    public int? DefaultDriverId { get; set; }
    public int? FuelCardCompanyId { get; set; }
    /// <summary>Required when a fuel card company is chosen. On one vehicle in the fleet at a time.</summary>
    public string? FuelCardNumber { get; set; }
    public int? TrackerCompanyId { get; set; }
    public string? TrackerDeviceId { get; set; }
    public string? Remarks { get; set; }

    /// <summary>Not in the future. Needs the acquisition permission; ignored for anyone without it.</summary>
    public DateOnly? AcquisitionDate { get; set; }
    /// <summary>Purchase, Lease, Rent, SharedInduction or CustomerInduction.</summary>
    public string? AcquisitionType { get; set; }

    /// <summary>The wizard's answers for steps 2 to 5 while the vehicle is a Draft, kept as they were sent so the wizard can be reopened (FR-VH-012). Up to 200 KB of JSON.</summary>
    public string? DraftData { get; set; }

    /// <summary>The fuel card number is on another vehicle in the fleet: say true to take it from that one (VAL-VH-018).</summary>
    public bool ReassignFuelCard { get; set; }
}

public class CreateVehicleRequest : VehicleInput;

public class UpdateVehicleRequest : VehicleInput
{
    /// <summary>The <c>rowVersion</c> the screen was loaded with. If the vehicle changed since, the save is refused.</summary>
    public string RowVersion { get; set; } = string.Empty;
}

public class PartnerRef
{
    public int Id { get; set; }
    public string BpCode { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
}

/// <summary>One dated ownership relation, current or past.</summary>
public class RelationModel
{
    public int Id { get; set; }
    public string Category { get; set; } = string.Empty;
    public PartnerRef? Counterparty { get; set; }
    public DateOnly EffectiveFrom { get; set; }
    public DateOnly? EffectiveTo { get; set; }
    public DateOnly? AgreementEndDate { get; set; }
    public string? AgreementReference { get; set; }
    public decimal? SharePercent { get; set; }
    public string? SharingBasis { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? FixedMonthlyAmount { get; set; }
    public string? ExpenseSharingRule { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? RentAmount { get; set; }
    public string? RentFrequency { get; set; }
    public int? RentDueDay { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? SecurityDeposit { get; set; }
    public string? ArrangementType { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? AgreedAmount { get; set; }
    public decimal? RevenueSharePercent { get; set; }
}

/// <summary>A vehicle as the screen shows it.</summary>
public class VehicleModel : VehicleInput
{
    public int Id { get; set; }
    public string VehicleCode { get; set; } = string.Empty;
    public string Status { get; set; } = string.Empty;
    public string? CurrentCategory { get; set; }
    public PartnerRef? CurrentCounterparty { get; set; }
    public PartnerRef? DefaultDriver { get; set; }
    public PartnerRef? FuelCardCompany { get; set; }
    public PartnerRef? TrackerCompany { get; set; }
    /// <summary>The open relation, or null for a Self Owned or Draft vehicle.</summary>
    public RelationModel? Relation { get; set; }
    /// <summary>Every relation the vehicle has had, newest first.</summary>
    public List<RelationModel> RelationHistory { get; set; } = [];
    public DateTime CreatedOn { get; set; }
    public DateTime ModifiedOn { get; set; }
    public string RowVersion { get; set; } = string.Empty;
}

// ── Category ──────────────────────────────────────────────────────────────────────

/// <summary>The fields of an ownership category (FSD §17). Only the ones of the chosen category are read.</summary>
public class CategoryDetails
{
    /// <summary>The bank, lessor, customer or sharing partner. Not used for Self Owned.</summary>
    public int? CounterpartyId { get; set; }
    /// <summary>When the arrangement starts. Not in the future.</summary>
    public DateOnly? StartDate { get; set; }
    public DateOnly? EndDate { get; set; }
    public string? AgreementReference { get; set; }

    public decimal? SharePercent { get; set; }
    /// <summary>ProfitShare, RevenueShare or FixedMonthly.</summary>
    public string? SharingBasis { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? FixedMonthlyAmount { get; set; }
    /// <summary>AllExpensesSameRatio, OnlyMajorExpenses or EachBearsOwn.</summary>
    public string? ExpenseSharingRule { get; set; }

    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? RentAmount { get; set; }
    /// <summary>Monthly, Weekly or PerTrip.</summary>
    public string? RentFrequency { get; set; }
    public int? RentDueDay { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? SecurityDeposit { get; set; }

    /// <summary>DedicatedMonthly, PerTrip or RevenueShare.</summary>
    public string? ArrangementType { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_FINANCE_VIEW)] public decimal? AgreedAmount { get; set; }
    public decimal? RevenueSharePercent { get; set; }
}

public class ChangeCategoryRequest
{
    /// <summary>SelfOwned, Shared, Rented, CustomerArrangement or BankLeased.</summary>
    public string Category { get; set; } = string.Empty;
    /// <summary>The day the new arrangement starts. The old one ends the day before. Not in the future.</summary>
    public DateOnly? EffectiveDate { get; set; }
    public string? Reason { get; set; }
    public CategoryDetails Details { get; set; } = new();
}

// ── Status, disposal ──────────────────────────────────────────────────────────────

public class ChangeStatusRequest
{
    /// <summary>Active, UnderMaintenance or TemporarilyUnavailable.</summary>
    public string Status { get; set; } = string.Empty;
    public string? Reason { get; set; }
    /// <summary>Defaults to today. Not in the future.</summary>
    public DateOnly? EffectiveDate { get; set; }
}

public class DisposeRequest
{
    /// <summary>Retire, Sell or Transfer.</summary>
    public string Kind { get; set; } = string.Empty;
    public DateOnly? Date { get; set; }
    public string? Reason { get; set; }
    /// <summary>The buyer, or the party the vehicle goes to. Required to sell or transfer.</summary>
    public int? CounterpartyId { get; set; }
    /// <summary>Required to sell.</summary>
    public decimal? Amount { get; set; }
    public string? Reference { get; set; }
}

public class LifecycleItem
{
    public string EventType { get; set; } = string.Empty;
    public string? FromStatus { get; set; }
    public string? ToStatus { get; set; }
    public string? FromCategory { get; set; }
    public string? ToCategory { get; set; }
    public DateOnly EffectiveDate { get; set; }
    public string? Reason { get; set; }
    public string? Reference { get; set; }
    public PartnerRef? Counterparty { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_COST_VIEW)] public decimal? Amount { get; set; }
    public string? UserName { get; set; }
    public DateTime OccurredOn { get; set; }
}

/// <summary>One change to a vehicle or to what belongs to it, from the audit trail.</summary>
public class VehicleHistoryChange
{
    public long Id { get; set; }
    public DateTime OccurredAt { get; set; }
    public Guid GroupId { get; set; }
    public string? UserName { get; set; }
    public string Entity { get; set; } = string.Empty;
    public string RecordId { get; set; } = string.Empty;
    public string Action { get; set; } = string.Empty;
    public string? Field { get; set; }
    public string? OldValue { get; set; }
    public string? NewValue { get; set; }
    public string? Reason { get; set; }
    public bool Restricted { get; set; }
}

public class VehicleHistory
{
    public PaginatedResponse<VehicleHistoryChange> Changes { get; set; } = new();
    public List<LifecycleItem> Lifecycle { get; set; } = [];
}

// ── Attached items ────────────────────────────────────────────────────────────────

public class AttachItemRequest
{
    /// <summary>A value of the Attached Item Type list.</summary>
    public int ItemTypeId { get; set; }
    public string Description { get; set; } = string.Empty;
    /// <summary>Unique among attached items. A container number for a container.</summary>
    public string? SerialNo { get; set; }
    /// <summary>A Vendor or Body Maker partner.</summary>
    public int? SupplierId { get; set; }
    /// <summary>Not before the vehicle's acquisition date, not in the future.</summary>
    public DateOnly? InstallationDate { get; set; }
    /// <summary>Needs the cost permission; ignored for anyone without it.</summary>
    public decimal? Cost { get; set; }
    public DateOnly? WarrantyUntil { get; set; }
    /// <summary>New, Used or Refurbished.</summary>
    public string? Condition { get; set; }
}

public class ItemModel
{
    public int Id { get; set; }
    public int VehicleId { get; set; }
    public int ItemTypeId { get; set; }
    public string? ItemType { get; set; }
    public string Description { get; set; } = string.Empty;
    public string? SerialNo { get; set; }
    public PartnerRef? Supplier { get; set; }
    public DateOnly InstallationDate { get; set; }
    [FieldPermission(PermissionCodes.VEH_FIELD_COST_VIEW)] public decimal? Cost { get; set; }
    public DateOnly? WarrantyUntil { get; set; }
    public string? Condition { get; set; }
    public string Status { get; set; } = string.Empty;
    public DateOnly? DetachedOn { get; set; }
    public string? DetachReason { get; set; }
    public int? TransferredFromItemId { get; set; }
    public int? TransferredToVehicleId { get; set; }
}

public class DetachItemRequest
{
    public DateOnly? Date { get; set; }
    public string? Reason { get; set; }
}

public class TransferItemRequest
{
    public int TargetVehicleId { get; set; }
    public DateOnly? Date { get; set; }
    public string? Reason { get; set; }
}

// ── Odometer, driver ──────────────────────────────────────────────────────────────

public class AddOdometerRequest
{
    public DateOnly? ReadingDate { get; set; }
    public int Km { get; set; }
    public string? Notes { get; set; }
}

public class OdometerModel
{
    public int Id { get; set; }
    public DateOnly ReadingDate { get; set; }
    public int Km { get; set; }
    public string Source { get; set; } = string.Empty;
    public string? Notes { get; set; }
}

public class AssignDriverRequest
{
    public int DriverId { get; set; }
    /// <summary>Defaults to today. Not in the future.</summary>
    public DateOnly? EffectiveDate { get; set; }
    /// <summary>The driver is the default driver of another vehicle: say true to release that assignment (VAL-VH-017).</summary>
    public bool ReleaseFromOther { get; set; }
}

public class ReleaseDriverRequest
{
    public DateOnly? Date { get; set; }
    public string? Reason { get; set; }
}

// ── List ──────────────────────────────────────────────────────────────────────────

public class VehicleListItem
{
    public int Id { get; set; }
    public string VehicleCode { get; set; } = string.Empty;
    public string RegistrationNo { get; set; } = string.Empty;
    public int VehicleTypeId { get; set; }
    public string? VehicleType { get; set; }
    public int MakeId { get; set; }
    public string? Make { get; set; }
    public string Model { get; set; } = string.Empty;
    public string? CurrentCategory { get; set; }
    public PartnerRef? Counterparty { get; set; }
    public Guid? BranchId { get; set; }
    public string? Branch { get; set; }
    public string Status { get; set; } = string.Empty;
    public PartnerRef? Driver { get; set; }
    public DateTime ModifiedOn { get; set; }
}

public class VehicleListQuery
{
    public int Page { get; set; } = 1;
    public int PageSize { get; set; } = 25;
    /// <summary>Matches registration number, chassis, engine and vehicle code: partial, from 3 characters.</summary>
    public string? Search { get; set; }
    /// <summary>modifiedOn (default, newest first), registrationNo, vehicleCode, status, category. Append <c>,asc</c> or <c>,desc</c>.</summary>
    public string? Sort { get; set; }
    public List<string>? Category { get; set; }
    public List<string>? Status { get; set; }
    public int? VehicleTypeId { get; set; }
    public int? MakeId { get; set; }
    public Guid? BranchId { get; set; }
    public int? DriverId { get; set; }
}

File: VMSFrontend/src/app/core/vehicle.models.ts
export interface PartnerRef {
  id: number;
  bpCode: string;
  name: string;
}

/** What is sent to save a vehicle's identity, technical and operational fields (FSD §16). */
export interface VehicleInput {
  registrationNo: string;
  registrationCityId?: number | null;
  chassisNo?: string | null;
  engineNo?: string | null;
  vehicleTypeId: number | null;
  makeId: number | null;
  model: string;
  manufacturingYear?: number | null;
  colour?: string | null;
  fuelType: string;
  tankCapacity?: number | null;
  loadCapacity?: number | null;
  capacityUnit?: string | null;
  axleConfigurationId?: number | null;
  bodyTypeId?: number | null;
  tyreCount?: number | null;
  gvw?: number | null;
  branchId?: string | null;
  openingOdometer?: number | null;
  openingOdometerDate?: string | null;
  defaultDriverId?: number | null;
  fuelCardCompanyId?: number | null;
  fuelCardNumber?: string | null;
  trackerCompanyId?: number | null;
  trackerDeviceId?: string | null;
  remarks?: string | null;
  acquisitionDate?: string | null;
  acquisitionType?: string | null;
  draftData?: string | null;
  reassignFuelCard?: boolean;
}

export interface UpdateVehicleRequest extends VehicleInput {
  rowVersion: string;
}

export interface Relation {
  id: number;
  category: string;
  counterparty?: PartnerRef | null;
  effectiveFrom: string;
  effectiveTo?: string | null;
  agreementEndDate?: string | null;
  agreementReference?: string | null;
  sharePercent?: number | null;
  sharingBasis?: string | null;
  fixedMonthlyAmount?: number | null;
  expenseSharingRule?: string | null;
  rentAmount?: number | null;
  rentFrequency?: string | null;
  rentDueDay?: number | null;
  securityDeposit?: number | null;
  arrangementType?: string | null;
  agreedAmount?: number | null;
  revenueSharePercent?: number | null;
}

export interface Vehicle extends VehicleInput {
  id: number;
  vehicleCode: string;
  status: string;
  currentCategory?: string | null;
  currentCounterparty?: PartnerRef | null;
  defaultDriver?: PartnerRef | null;
  fuelCardCompany?: PartnerRef | null;
  trackerCompany?: PartnerRef | null;
  relation?: Relation | null;
  relationHistory: Relation[];
  createdOn: string;
  modifiedOn: string;
  rowVersion: string;
}

export interface VehicleListItem {
  id: number;
  vehicleCode: string;
  registrationNo: string;
  vehicleTypeId: number;
  vehicleType?: string | null;
  makeId: number;
  make?: string | null;
  model: string;
  currentCategory?: string | null;
  counterparty?: PartnerRef | null;
  branchId?: string | null;
  branch?: string | null;
  status: string;
  driver?: PartnerRef | null;
  modifiedOn: string;
}

export interface CategoryDetails {
  counterpartyId?: number | null;
  endDate?: string | null;
  agreementReference?: string | null;
  sharePercent?: number | null;
  sharingBasis?: string | null;
  fixedMonthlyAmount?: number | null;
  expenseSharingRule?: string | null;
  rentAmount?: number | null;
  rentFrequency?: string | null;
  rentDueDay?: number | null;
  securityDeposit?: number | null;
  arrangementType?: string | null;
  agreedAmount?: number | null;
  revenueSharePercent?: number | null;
}

export interface ChangeCategoryRequest {
  category: string;
  effectiveDate?: string | null;
  reason?: string | null;
  details: CategoryDetails;
}

export interface DisposeRequest {
  kind: string;
  date?: string | null;
  reason?: string | null;
  counterpartyId?: number | null;
  amount?: number | null;
  reference?: string | null;
}

export interface AttachedItem {
  id: number;
  vehicleId: number;
  itemTypeId: number;
  itemType?: string | null;
  description: string;
  serialNo?: string | null;
  supplier?: PartnerRef | null;
  installationDate: string;
  /** Absent when the signed-in user may not see costs. */
  cost?: number | null;
  warrantyUntil?: string | null;
  condition?: string | null;
  status: string;
  detachedOn?: string | null;
  detachReason?: string | null;
  transferredFromItemId?: number | null;
  transferredToVehicleId?: number | null;
}

export interface AttachItemRequest {
  itemTypeId: number | null;
  description: string;
  serialNo?: string | null;
  supplierId?: number | null;
  installationDate?: string | null;
  cost?: number | null;
  warrantyUntil?: string | null;
  condition?: string | null;
}

export interface OdometerReading {
  id: number;
  readingDate: string;
  km: number;
  source: string;
  notes?: string | null;
}

export interface LifecycleItem {
  eventType: string;
  fromStatus?: string | null;
  toStatus?: string | null;
  fromCategory?: string | null;
  toCategory?: string | null;
  effectiveDate: string;
  reason?: string | null;
  reference?: string | null;
  counterparty?: PartnerRef | null;
  amount?: number | null;
  userName?: string | null;
  occurredOn: string;
}

export interface VehicleHistoryChange {
  id: number;
  occurredAt: string;
  groupId: string;
  userName?: string | null;
  entity: string;
  recordId: string;
  action: string;
  field?: string | null;
  oldValue?: string | null;
  newValue?: string | null;
  reason?: string | null;
  restricted: boolean;
}

export interface VehicleHistory {
  changes: { items: VehicleHistoryChange[]; totalCount: number; page: number; pageSize: number; totalPages: number };
  lifecycle: LifecycleItem[];
}

/** A way a vehicle is linked to a partner. */
export interface LinkedVehicle {
  vehicleId: number;
  vehicleCode: string;
  registrationNo: string;
  vehicleStatus: string;
  link: string;
  detail?: string | null;
  from?: string | null;
  to?: string | null;
  isCurrent: boolean;
}

// ── Acquisition and finance (FSD §18) ──────────────────────────────────────────────

/** The acquisition block of a Draft. Amounts are absent for someone without `VEH.FIELD.COST.VIEW`. */
export interface Acquisition {
  vehicleId: number;
  vehicleStatus: string;
  acquisitionDate?: string | null;
  acquisitionType?: string | null;
  sellerId?: number | null;
  seller?: PartnerRef | null;
  purchasePrice?: number | null;
  amountPaid?: number | null;
  paymentMode?: string | null;
  paymentReference?: string | null;
  registrationCost?: number | null;
  /** The vehicle's row version: sent back so a change made by someone else in the meantime is not overwritten. */
  rowVersion: string;
}

export interface SaveAcquisitionRequest {
  acquisitionDate: string | null;
  acquisitionType: string | null;
  sellerId: number | null;
  purchasePrice: number | null;
  amountPaid: number | null;
  paymentMode: string | null;
  paymentReference: string | null;
  registrationCost: number | null;
  rowVersion: string;
}

/** The bank's agreement for a vehicle. Amounts are absent for someone without `VEH.FIELD.FINANCE.VIEW`. */
export interface Agreement {
  id: number;
  vehicleId: number;
  financeTypeId: number;
  financeType?: string | null;
  bankId: number;
  bank?: PartnerRef | null;
  agreementNo: string;
  agreementDate: string;
  financeAmount?: number;
  downPayment?: number;
  installmentAmount?: number;
  frequency: string;
  tenure: number;
  firstDueDate: string;
  markupRate?: number | null;
  residualAmount?: number | null;
  securityDeposit?: number | null;
  totalPayable?: number;
  status: string;
  rowVersion: string;
}

export interface SaveFinanceRequest {
  financeTypeId: number | null;
  bankId: number | null;
  agreementNo: string;
  agreementDate: string | null;
  financeAmount: number | null;
  downPayment: number | null;
  installmentAmount: number | null;
  frequency: string;
  tenure: number | null;
  firstDueDate: string | null;
  markupRate: number | null;
  residualAmount: number | null;
  securityDeposit: number | null;
  /** Set to go ahead when finance plus down payment is not the purchase price. */
  confirmMismatch: boolean;
  rowVersion?: string | null;
}
// ── Activation, the summary, the schedule and the ledger ───────────────────────────

/** The fields of an ownership category as the API takes them (FSD §17). Amounts need `VEH.FIELD.FINANCE.VIEW`. */
export interface CategoryDetails {
  counterpartyId?: number | null;
  endDate?: string | null;
  agreementReference?: string | null;
  sharePercent?: number | null;
  sharingBasis?: string | null;
  fixedMonthlyAmount?: number | null;
  expenseSharingRule?: string | null;
  rentAmount?: number | null;
  rentFrequency?: string | null;
  rentDueDay?: number | null;
  securityDeposit?: number | null;
  arrangementType?: string | null;
  agreedAmount?: number | null;
  revenueSharePercent?: number | null;
}

export interface ScheduleDueDate {
  installmentNo: number;
  dueDate: string;
}

export interface ActivateVehicleRequest {
  category: string;
  details: CategoryDetails;
  items: AttachItemRequest[];
  dueDates: ScheduleDueDate[];
  reassignFuelCard: boolean;
  releaseFromOther: boolean;
  rowVersion?: string | null;
}

export interface ChecklistItem {
  code: string;
  label: string;
  ok: boolean;
  /** False for something worth knowing that does not stop the activation. */
  blocking: boolean;
  message?: string | null;
}

export interface PostingPreview {
  type: string;
  subType?: string | null;
  amount?: number;
  date: string;
  partner?: PartnerRef | null;
  reference: string;
}

export interface ScheduleRow {
  installmentNo: number;
  dueDate: string;
  amount?: number;
  isResidual: boolean;
}

export interface ActivationCheck {
  canActivate: boolean;
  items: ChecklistItem[];
  postings: PostingPreview[];
  schedule: ScheduleRow[];
}

export interface AgreementSummary {
  agreementId: number;
  status: string;
  bank?: PartnerRef | null;
  totalPayable?: number;
  residual?: number;
  outstanding?: number;
  installmentsPaid: number;
  installmentsTotal: number;
  nextDueDate?: string | null;
  nextDueAmount?: number | null;
  overdueCount: number;
  overdueAmount?: number;
}

export interface FinancialSummary {
  vehicleId: number;
  acquisitionCost?: number;
  majorExpenses?: number;
  totalCost?: number;
  paidToDate?: number;
  deposits?: number;
  agreement?: AgreementSummary | null;
}

export interface Installment {
  id: number;
  agreementId: number;
  installmentNo: number;
  dueDate: string;
  expectedAmount?: number;
  paidAmount?: number;
  remainingAmount?: number;
  paidOn?: string | null;
  isResidual: boolean;
  status: string;
  rowVersion: string;
}

export interface PayInstallmentRequest {
  amount: number | null;
  paidOn: string | null;
  paymentMode: string | null;
  reference: string | null;
  rowVersion: string;
}

export interface InstallmentPayment {
  installment: Installment;
  transactionId: number;
  agreementSettled: boolean;
}

// ── Recurring charges and payables (FSD §19A) ───────────────────────────────────────

/** One row of a vehicle's configured recurring charges, current or historic (BR-VH-033). */
export interface RecurringCharge {
  id: number;
  seriesId: string;
  vehicleId: number;
  chargeTypeId: number;
  chargeType?: string | null;
  payee?: PartnerRef | null;
  expenseTypeId: number;
  expenseType?: string | null;
  amount?: number | null;
  amountBasis: string;
  frequency: string;
  dueDay?: number | null;
  dueMonth?: number | null;
  customIntervalDays?: number | null;
  startDate: string;
  endDate?: string | null;
  occurrenceCount?: number | null;
  generatedCount: number;
  postingMode: string;
  generateLeadDays: number;
  taxWithholdingPercent?: number | null;
  nextDueDate?: string | null;
  isActive: boolean;
  effectiveTo?: string | null;
  endReason?: string | null;
  rowVersion: string;
}

export interface SaveRecurringChargeRequest {
  chargeTypeId: number | null;
  payeeId: number | null;
  expenseTypeId: number | null;
  amount: number | null;
  amountBasis: string | null;
  frequency: string | null;
  dueDay?: number | null;
  dueMonth?: number | null;
  customIntervalDays?: number | null;
  startDate: string | null;
  endDate?: string | null;
  occurrenceCount?: number | null;
  postingMode: string | null;
  generateLeadDays?: number | null;
  taxWithholdingPercent?: number | null;
  rowVersion?: string | null;
}

export interface EndRecurringChargeRequest {
  endDate?: string | null;
  reason: string | null;
}

/** One Due or Overdue item, of either kind (BR-VH-029): a generated charge entry, or an installment surfaced without a parallel schedule. */
export interface Payable {
  id: number;
  kind: 'RecurringCharge' | 'Installment';
  vehicleId: number;
  vehicleRegistrationNo?: string | null;
  branchId?: string | null;
  branch?: string | null;
  chargeId?: number | null;
  chargeTypeId?: number | null;
  chargeType?: string | null;
  payee?: PartnerRef | null;
  dueDate: string;
  expectedAmount?: number | null;
  status: string;
}

export interface ConfirmChargeEntryRequest {
  amount: number | null;
  paidOn: string | null;
  paymentMode?: string | null;
  reference?: string | null;
  remarks?: string | null;
}

export interface WaiveChargeEntryRequest {
  reason: string | null;
}

export interface ChargeEntryPayment {
  transactionId: number;
}

export interface BulkConfirmItem {
  kind: 'RecurringCharge' | 'Installment';
  vehicleId: number;
  id: number;
  rowVersion?: string | null;
  amount?: number | null;
}

export interface BulkConfirmRequest {
  items: BulkConfirmItem[];
  paidOn: string | null;
  paymentMode?: string | null;
}

export interface BulkConfirmFailure {
  kind: string;
  vehicleId: number;
  id: number;
  message: string;
}

export interface BulkConfirmResult {
  succeeded: number;
  failed: BulkConfirmFailure[];
}

export interface PayablesSummary {
  dueWithin7Days: number;
  dueWithin7DaysAmount?: number;
  overdueCount: number;
  overdueAmount?: number;
}

export interface LedgerEntry {
  id: number;
  type: string;
  subType?: string | null;
  amount?: number;
  date: string;
  partner?: PartnerRef | null;
  reference?: string | null;
  source: string;
  isSystemGenerated: boolean;
  reason?: string | null;
  reversesTransactionId?: number | null;
  isReversed: boolean;
  /** A receipt was uploaded for this entry (source §7). */
  hasReceipt: boolean;
  receiptFileName?: string | null;
  createdOn: string;
}
